src/nvim/lua/executor.c

```diff
@@ -1008,3 +1012,244 @@ static void nlua_add_treesitter(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
+  return 0;
+}
+
+// Stores the header field name on LUA_HTTP_LAST_HEADER_FIELD so that the
+// header_value callback can use it to set the header value on the
+// metatable.headers table
+static int nlua_http_parser_on_header_field(http_parser *p,
+                                            const char *at,
+                                            size_t length)
+{
+  lua_State *lstate = nlua_enter();
+  lua_getfield(lstate, -1, LUA_HTTP_HEADERS_KEY);  // [meta, headers]
+  lua_pushlstring(lstate, at, length);             // [meta, headers, value]
+  lua_setfield(lstate, -2, LUA_HTTP_LAST_HEADER_FIELD);  // [meta, headers]
+  lua_pop(lstate, 1);                                    // [meta]
+  return 0;
+}
+
+static int nlua_http_parser_on_header_value(http_parser *p,
+                                            const char *at,
+                                            size_t length)
+{
+  lua_State *lstate = nlua_enter();
+  lua_getfield(lstate, -1, LUA_HTTP_HEADERS_KEY);        // [meta, headers]
+  lua_getfield(lstate, -1, LUA_HTTP_LAST_HEADER_FIELD);  // [meta, headers, key]
+  lua_pushlstring(lstate, at, length);  // [meta, headers, key, value]
+  lua_rawset(lstate, -3);               // [meta, headers]
+  // TODO(ashkan): nil out LUA_HTTP_LAST_HEADER_FIELD when I'm done here?
+  lua_pop(lstate, 1);  // [meta]
+  return 0;
+}
+
+http_parser_settings lua_http_parser_settings = {
+    .on_url = nlua_http_parser_on_url,
+    .on_body = nlua_http_parser_on_body,
+    .on_header_field = nlua_http_parser_on_header_field,
+    .on_header_value = nlua_http_parser_on_header_value,
+    .on_headers_complete = nlua_http_parser_on_complete_headers,
+    .on_message_complete = nlua_http_parser_on_complete_message,
+};
+
+static int nlua_http_parser_execute(lua_State *const lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  http_parser *p = luaL_checkudata(lstate, 1, "http_parser");
+  size_t chunk_len = 0;
+  const char *chunk = luaL_checklstring(lstate, 2, &chunk_len);
+  lua_getfenv(lstate, 1);
+  size_t bytes_parsed
+      = http_parser_execute(p, &lua_http_parser_settings, chunk, chunk_len);
+  lua_pop(lstate, 1);
+  if (p->http_errno) {
+    // TODO(ashkan): use the http_errno_name() somehow?
+    lua_pushstring(lstate, http_errno_description(p->http_errno));
+    return lua_error(lstate);
+  }
+
+  lua_Integer ret = (lua_Integer)bytes_parsed;
+  lua_pushinteger(lstate, ret);
+
+  // TODO(ashkan): do something if this is finished?
+  // if (http_body_is_final(p)) {
+  //   lua_getfenv(lstate, 1);
+  //   return 2;
+  // }
+
+  return 1;
+}
+
+// TODO(ashkan): improve diagnostics?
+static int nlua_http_parser_to_string(lua_State *const lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  lua_pushliteral(lstate, "<http_parser>");
+  return 1;
+}
+
+static int nlua_http_parser_index(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
+{
+  // TODO(ashkan): check this even for index?
+  // http_parser* p = luaL_checkudata(lstate, 1, "http_parser");
```

Comment:
This check is basically unnecessary if being used on __index but if some sick person decided to try to use the http_parser metatable elsewhere, the check would be useful.

Reply:
checks are good...